Pcap++/header/DeviceListBase.h

```diff
@@ -0,0 +1,167 @@
+#pragma once
+
+/// @file
+
+#include "PointerVector.h"
+
+namespace pcpp
+{
+	namespace internal
+	{
+		/**
+		 * @class DeviceListBase
+		 * A base class for all device lists in PcapPlusPlus. This class is used to store a list of devices and provide
+		 * access to them
+		 */
+		template <class DeviceType, class DeviceDeleter = std::default_delete<DeviceType>> class DeviceListBase
```

Comment:
I would prefer keeping it separate from `PointerVector` as an abstraction layer. 

Changing it to a generic `ReadOnlyPointerVector` would mean that it can't keep the protected destructor (as that may be used in other places) and will have to have a virtual public destructor to not cause issues with deletion from base pointer.

Also, it might allow centralising other shared functions between the different lists if there are any to be had.

Reply:
In order to support a protected d'tor we can create `ReadOnlyPointerVector` with a non-protected destructor, and then inherit from it and add a protected d'tor, or just add a protected d'tor to all list classes

Reply:
> In order to support a protected d'tor we can create ReadOnlyPointerVector with a non-protected destructor, and then inherit from it and add a protected d'tor, or just add a protected d'tor to all list classes.

That does not solve the problem tho... the protected destructor is to prevent accidental deletion from a base class pointer, as `DeviceListBase`'s inheritance is public and avoiding the need for having `virtual` destructor which adds runtime overhead. Having a base class above `DeviceListBase` with a public destructor just pushes the problem up the chain, as now code like this is valid:
```
ReadOnlyPointerVector<T>* p = new DerivedFromDeviceBaseList();
delete p; // Calls ~ReadOnlyPointerVector(). Does not destruct anything down the inheritance chain.
```

This can be solved by inheriting privately from a potential `ReadOnlyPointerVector` but really do we even need a general purpose `ReadOnlyPointerVector` at the moment? What would be the difference between that and using `PointerVector<T> const&` where it is needed. I'm of the opinion to leave it be for now and add a `ReadOnlyPointerVector` if another place actually requires it.